# Artificial/Deep Neural Network

In [1]:
import numpy as np
import pandas as pd
import tensorflow as tf
tf.__version__

'2.9.2'

In [2]:
dataset = pd.read_csv('/content/Churn_Modelling.csv')
dataset.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [3]:
# split the data into dependent and independent variable
x = dataset.iloc[:,3:-1]
y = dataset.iloc[:,-1]

In [4]:
x.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

In [5]:
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [6]:
x = pd.get_dummies(x, columns=['Geography',	'Gender'])
x

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42       2       0.00              1          1   
1             608   41       1   83807.86              1          0   
2             502   42       8  159660.80              3          1   
3             699   39       1       0.00              2          0   
4             850   43       2  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39       5       0.00              2          1   
9996          516   35      10   57369.61              1          1   
9997          709   36       7       0.00              1          0   
9998          772   42       3   75075.31              2          1   
9999          792   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Geography_France  Geography_Germany  \
0                  1        101348.88                 1                  0   
1                  1        112542.58                 0                  0   
2                  0        113931.57                 1                  0   
3                  0         93826.63                 1                  0   
4                  1         79084.10                 0                  0   
...              ...              ...               ...                ...   
9995               0         96270.64                 1                  0   
9996               1        101699.77                 1                  0   
9997               1         42085.58                 1                  0   
9998               0         92888.52                 0                  1   
9999               0         38190.78                 1                  0   

      Geography_Spain  Gender_Female  Gender_Male  
0                   0              1            0  
1                   1              1            0  
2                   0              1            0  
3                   0              1            0  
4                   1              1            0  
...               ...            ...          ...  
9995                0              0            1  
9996                0              0            1  
9997                0              1            0  
9998                0              0            1  
9999                0              1            0  

[10000 rows x 13 columns]

In [7]:
x = x.drop(['Geography_France','Gender_Female'], axis=1 )

In [8]:
x.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
0               1        101348.88                  0                0   
1               1        112542.58                  0                1   
2               0        113931.57                  0                0   
3               0         93826.63                  0                0   
4               1         79084.10                  0                1   

   Gender_Male  
0            0  
1            0  
2            0  
3            0  
4            0

In [14]:
x.shape

(10000, 11)

In [10]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_stand = sc.fit_transform(x)
x_stand

array([[-0.32622142,  0.29351742, -1.04175968, ..., -0.57873591,
        -0.57380915, -1.09598752],
       [-0.44003595,  0.19816383, -1.38753759, ..., -0.57873591,
         1.74273971, -1.09598752],
       [-1.53679418,  0.29351742,  1.03290776, ..., -0.57873591,
        -0.57380915, -1.09598752],
       ...,
       [ 0.60498839, -0.27860412,  0.68712986, ..., -0.57873591,
        -0.57380915, -1.09598752],
       [ 1.25683526,  0.29351742, -0.69598177, ...,  1.72790383,
        -0.57380915,  0.91241915],
       [ 1.46377078, -1.04143285, -0.35020386, ..., -0.57873591,
        -0.57380915, -1.09598752]])

In [12]:
# split the data into training and test
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x_stand, y, test_size=0.2, random_state=1)

# Building Deep Neural Network (MultiLayerPerceptron)

In [13]:
from tensorflow import keras

In [15]:
# Create API
dnn = tf.keras.models.Sequential()
# Adding first hidden Layer
dnn.add(tf.keras.layers.Dense(units=6, activation='relu'))

# Adding second hidden Layer
dnn.add(tf.keras.layers.Dense(units=6, activation='relu'))

# adding output
dnn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Train the model
dnn.compile(optimizer='adam', loss = 'binary_crossentropy', metrics =['accuracy'])

# apply training dataset and fit the model
dnn.fit(x_train, y_train, batch_size=32, epochs=100)

Epoch 1/100
250/250 [==============================] - 3s 2ms/step - loss: 0.5636 - accuracy: 0.7439
Epoch 2/100
250/250 [==============================] - 1s 2ms/step - loss: 0.4697 - accuracy: 0.7997
Epoch 3/100
250/250 [==============================] - 1s 2ms/step - loss: 0.4423 - accuracy: 0.8069
Epoch 4/100
250/250 [==============================] - 1s 3ms/step - loss: 0.4290 - accuracy: 0.8142
Epoch 5/100
250/250 [==============================] - 1s 3ms/step - loss: 0.4219 - accuracy: 0.8185
Epoch 6/100
250/250 [==============================] - 1s 3ms/step - loss: 0.4172 - accuracy: 0.8224
Epoch 7/100
250/250 [==============================] - 1s 2ms/step - loss: 0.4140 - accuracy: 0.8223
Epoch 8/100
250/250 [==============================] - 1s 2ms/step - loss: 0.4117 - accuracy: 0.8251
Epoch 9/100
250/250 [==============================] - 1s 2ms/step - loss: 0.4100 - accuracy: 0.8274
Epoch 10/100
250/250 [==============================] - 1s 2ms/step - loss: 0.4085 - accura

In [16]:
dnn.fit(x_train, y_train, batch_size=32, epochs=100)

Epoch 1/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3322 - accuracy: 0.8631
Epoch 2/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3320 - accuracy: 0.8614
Epoch 3/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3321 - accuracy: 0.8622
Epoch 4/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3324 - accuracy: 0.8622
Epoch 5/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3323 - accuracy: 0.8624
Epoch 6/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3319 - accuracy: 0.8624
Epoch 7/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3314 - accuracy: 0.8626
Epoch 8/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3315 - accuracy: 0.8634
Epoch 9/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3317 - accuracy: 0.8622
Epoch 10/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3318 - accura

In [25]:
dnn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 6)                 72        
                                                                 
 dense_1 (Dense)             (None, 6)                 42        
                                                                 
 dense_2 (Dense)             (None, 1)                 7         
                                                                 
Total params: 121
Trainable params: 121
Non-trainable params: 0
_________________________________________________________________


### First Layer (`dense`)
- **Output shape**: `(None, 6)` — This means the layer outputs 6 neurons, with `None` representing the batch size (flexible input).
- **Parameters**: 72 — This suggests an input size of 11 (since \( 11 \times 6 + 6 = 66 + 6 = 72 \), accounting for weights and biases).

### Second Layer (`dense_1`)
- **Output shape**: `(None, 6)` — Another layer with 6 neurons.
- **Parameters**: 42 — This is calculated as \( 6 \times 6 + 6 = 36 + 6 = 42 \), indicating the input from the previous layer (6 neurons) plus biases.

### Third Layer (`dense_2`)
- **Output shape**: `(None, 1)` — A single output neuron, likely for a binary classification or regression task.
- **Parameters**: 7 — Calculated as \( 6 \times 1 + 1 = 6 + 1 = 7 \), based on the 6 inputs from the previous layer plus one bias.

### Total Parameters
- **121** (sum of all layer parameters: \( 72 + 42 + 7 \)).

In [17]:
# Predict the test result
y_pred = dnn.predict(x_test)

63/63 [==============================] - 0s 1ms/step


In [19]:
y_pred = (y_pred>0.5)

In [20]:
from sklearn.metrics import accuracy_score, classification_report

In [22]:
accuracy_score(y_test, y_pred)

0.8655

In [24]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.97      0.92      1585
           1       0.79      0.48      0.60       415

    accuracy                           0.87      2000
   macro avg       0.83      0.72      0.76      2000
weighted avg       0.86      0.87      0.85      2000



**Class Imbalance:**

The dataset is heavily skewed toward class 0 (1585 vs. 415 samples). This imbalance likely causes the model to favor class 0, as seen in the high recall (0.97) and precision (0.88) for class 0, but poor recall (0.48) for class 1.

**Strengths:** 
The model is very good at identifying class 0 (high precision and recall), meaning it rarely misses class 0 instances and is confident when it predicts class 0.

**Weaknesses:**

The model struggles with class 1, with a low recall (0.48), meaning it misses more than half of the actual class 1 instances (false negatives). Precision for class 1 (0.79) is decent but not as strong as for class 0.

**Overall Performance:**

The accuracy (87%) looks good, but it’s inflated by the imbalance. The macro avg F1-score (0.76) gives a more balanced view, showing the model isn’t equally effective for both classes.